src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj

```diff
@@ -206,126 +93,146 @@
+  (-> jruby-puppet-config
+      (assoc :http-client-ssl-protocols (:ssl-protocols http-config))
+      (assoc :http-client-cipher-suites (:cipher-suites http-config))
+      (assoc :http-client-connect-timeout-milliseconds
+             (get http-config :connect-timeout-milliseconds
+                  default-http-connect-timeout))
+      (assoc :http-client-idle-timeout-milliseconds
+             (get http-config :idle-timeout-milliseconds
+                  default-http-socket-timeout))
+      (update-in [:master-conf-dir] #(or % default-master-conf-dir))
+      (update-in [:master-var-dir] #(or % default-master-var-dir))
+      (update-in [:master-code-dir] #(or % default-master-code-dir))
+      (update-in [:master-run-dir] #(or % default-master-run-dir))
+      (update-in [:master-log-dir] #(or % default-master-log-dir))
+      (update-in [:max-requests-per-instance] #(or % 0))
+      (update-in [:use-legacy-auth-conf] #(or % (nil? %)))
+      (dissoc :environment-class-cache-enabled)))
 
-(schema/defn ^:always-validate
-  lock-pool
-  "Locks the JRuby pool for exclusive access."
-  [pool :- jruby-schemas/pool-queue-type
-   reason :- schema/Any
-   event-callbacks :- [IFn]]
-  (log/debug "Acquiring lock on JRubyPool...")
-  (lock-requested event-callbacks reason)
-  (.lock pool)
-  (lock-acquired event-callbacks reason)
-  (log/debug "Lock acquired"))
+(schema/defn create-jruby-config :- jruby-schemas/JRubyConfig
+  "Handles creating a valid JRubyConfig map for use in the jruby-puppet-service.
+  This method:
+  * Creates the appropriate lifecycle functions
+  * overrides the default ruby-load-path to include the ruby code directory from
+    this project"
+  [jruby-puppet-config :- jruby-puppet-schemas/JRubyPuppetConfig
+   jruby-config :- {schema/Keyword schema/Any}
+   agent-shutdown-fn :- IFn
+   profiler :- (schema/maybe PuppetProfiler)]
+  (let [initialize-pool-instance-fn (get-initialize-pool-instance-fn jruby-puppet-config profiler)
+        lifecycle-fns {:shutdown-on-error agent-shutdown-fn
+                       :initialize-pool-instance initialize-pool-instance-fn
+                       :cleanup cleanup-fn}
+        modified-jruby-config (assoc jruby-config :ruby-load-path (jruby-puppet-core/managed-load-path
+                                                                            (:ruby-load-path jruby-config)))]
+    (jruby-core/initialize-config (assoc modified-jruby-config :lifecycle lifecycle-fns))))
+
+(schema/defn initialize-and-create-jruby-config :- jruby-schemas/JRubyConfig
+  "Handles the initialization of the jruby-puppet config for the purpose of returning a
+  jruby config"
+  ([raw-config :- {schema/Keyword schema/Any}]
+   (initialize-and-create-jruby-config raw-config nil (fn [])))
+  ([raw-config :- {schema/Keyword schema/Any}
+    profiler :- (schema/maybe PuppetProfiler)
+    agent-shutdown-fn :- IFn]
+   (initialize-and-create-jruby-config raw-config profiler agent-shutdown-fn true))
```

Comment:
Yeah I'll switch it around. Probably keep the 1-arity version call the 4 arity version with false, and then have the service's call to this function pass in true, and get rid of the 3-arity